pkg/ingester/transfer.go

```diff
@@ -291,3 +400,396 @@ func (i *Ingester) findTargetIngester(ctx context.Context) (*ring.IngesterDesc,
+	// it's OK to shut down.
+	if err := stream.SendAndClose(&client.SendChunksResponse{}); err != nil {
+		level.Error(util.Logger).Log("msg", "Error closing SendChunks stream", "from_ingester", fromIngesterID, "err", err)
+		return err
+	}
+
+	// It's valid for an ingester not to have any tokens to send in a range,
+	// just ignore it here.
+	if seriesReceived == 0 {
+		return nil
+	}
+
+	if fromIngesterID == "" {
+		level.Error(util.Logger).Log("msg", "received TransferChunks request with no ID from ingester")
+		return fmt.Errorf("no ingester id")
+	}
+
+	level.Debug(util.Logger).Log("msg", "Successfully received chunks", "from_ingester", fromIngesterID, "series_received", seriesReceived)
+	return nil
+}
+
+// GetChunks accepts a get request from a client and sends all chunks from the serving ingester
+// that fall within the given range to the client.
+func (i *Ingester) GetChunks(req *client.GetChunksRequest, stream client.Ingester_GetChunksServer) error {
+	userStatesCopy := i.userStates.cp()
+	if len(userStatesCopy) == 0 {
+		level.Info(util.Logger).Log("msg", "nothing to transfer")
+		return nil
+	}
+
+	sent, err := i.pushChunks(pushChunksOptions{
+		States:        userStatesCopy,
+		Stream:        stream,
+		DisallowFlush: req.Move,
+		SentChunks:    incSentChunks,
+		SentSeries:    incSentSeries,
+
+		Filter: func(pair *fingerprintSeriesPair) bool {
+			token := pair.series.token
+
+			inRange := false
+			for _, rg := range req.Ranges {
+				if token >= rg.StartRange && token < rg.EndRange {
+					inRange = true
+					break
+				}
+			}
+			return !inRange
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	level.Debug(util.Logger).Log(
+		"msg", "sent chunks in range",
+		"to_ingester", req.FromIngesterId,
+		"ranges", ring.PrintableRanges(makeRingRanges(req.Ranges)),
+		"sent_chunks", sent,
+	)
+	return nil
+}
+
+// BlockTokenRange handles a remote request for a token range to be blocked.
+// Blocked ranges will automatically be unblocked after the RangeBlockPeriod
+// configuration variable.
+//
+// When a range is blocked, the Ingester will no longer accept pushes
+// for any streams whose tokens falls within the blocked ranges. Unblocking
+// the range re-enables those pushes.
+func (i *Ingester) BlockTokenRange(ctx context.Context, req *client.RangeRequest) (*client.BlockRangeResponse, error) {
+	i.blockedTokenMtx.Lock()
+	defer i.blockedTokenMtx.Unlock()
+
+	for _, rg := range req.Ranges {
+		for _, r := range i.blockedTokens {
+			if r.From == rg.StartRange && r.To == rg.EndRange {
+				// Multiple clients may request a block simultaneously. We silently
+				// exit so they can continue processing as normal.
+				level.Warn(util.Logger).Log("msg", "token range already blocked",
+					"start_token", rg.StartRange, "end_token", rg.EndRange)
+
+				continue
+			}
+		}
+
+		i.blockedTokens = append(i.blockedTokens, ring.TokenRange{
+			From: rg.StartRange,
+			To:   rg.EndRange,
+		})
+
+		blockedRanges.Inc()
+	}
```

Comment:
This may be simpler if we just have blockedTokens be a map with the key being the range.